Changed Application to Fleet (#36)
balena-io/balena-cli-advanced-masterclass: README.md

## Introduction

The balena Command Line Interface (balena CLI) utility consists of a number

> of commands that allow a user to develop, deploy and manage balena applications, Fleets, 

Reviewer: ```suggestion
of commands that allow a user to develop, deploy and manage balena fleets, 
```

### 1. Communicating with Alternative balena Environments

By default, the balena CLI communicates with the production balenaCloud instance,

> using this environment to carry out operations such as Fleet creation,

Reviewer: ```suggestion
using this environment to carry out operations such as fleet creation,
```

```

We'll be using the two separate environments (staging and production) in the

> next set of exercises to show you how devices can be moved between Fleets

Reviewer: ```suggestion
next set of exercises to show you how devices can be moved between fleets
```

and environments. Don't delete them just yet!

### 2. Moving Devices between Fleets and Environments

Usually you'll provision a device that will exist on a particular balena

> Fleet or environment, as the lifecycle of that device will only make

Reviewer: ```suggestion
fleet or environment, as the lifecycle of that device, will only make
```

> sense within that Fleet.

Reviewer: ```suggestion
sense within that fleet.
```

However, there are times where it is useful to be able to move a device either

> from one Fleet to another (for example when a major rewrite of your

Reviewer: ```suggestion
from one fleet to another (for example when a major rewrite of your
```

application occurs that is no longer backwards compatible with a prior version)
or from one environment to another (perhaps you've created a locally hosted
test environment using openBalena and now want to move from your test
environment to the production environment of balenaCloud).

The following exercises will show you how to carry this out.

#### 2.1 Moving Devices between Fleets

> Moving a device to another Fleet in the same environment is extremely
> easy. To demonstrate this, first create a new Fleet for the balenaFin:

Reviewer: ```suggestion
Moving a device to another fleet in the same environment is extremely
easy. To demonstrate this, first create a new fleet for the balenaFin:
```

```shell
$ cd ~/balenaProduction

> $ balena app create altFleet --type fincm3

Reviewer: ```suggestion
$ balena fleet create altFleet --type fincm3
```

Application created: altFleet (fincm3, id 987654)
```

We should already have a device connecting to our previous 'cliFleet' (from the
previous
[balena CLI Masterclass](https://github.com/balena-io/balena-cli-masterclass)). See the
[Creating an Application and Provisioning a Device](https://github.com/balena-io/balena-cli-masterclass#2-creating-an-application-and-provisioning-a-device)
section to create an application and provision a device against it, if you
haven't already done so, and wish to follow this exercise. Once a device
is provisioned against the `cliFleet` application and online, execute the
following command:

```shell
$ cd ~/balenaProduction
$ balena devices

> ID      UUID    DEVICE NAME          DEVICE TYPE     APPLICATION NAME STATUS IS ONLINE

Reviewer: ```suggestion
ID      UUID    DEVICE NAME          DEVICE TYPE     FLEET NAME STATUS IS ONLINE
```

1234556 7654321 weathered-wildflower fincm3          cliFleet           Idle   false     9.15.7             balenaOS 2.38.0+rev1 https://dashboard.balena-cloud.io/devices/1234567890abcdef/summary
```

> To interactively determine which Fleet to move a device to, simply use

Reviewer: ```suggestion
To interactively determine which fleet to move a device to, simply use
```

its UUID with the `balena device move` command:

```shell
$ cd ~/balenaProduction
$ balena device move 7654321
? Select an application (Use arrow keys)
❯ altFleet (fincm3)
  anotherFleet (raspberry-pi)
```

> As you can see, only Fleets that support the device type of the device that is being moved are available. For non-interactive movement, simply pass the

Reviewer: ```suggestion
As you can see, only fleets that support the device type of the device that is being moved are available. For non-interactive movement, simply pass the
```

> optional `--application` switch to the command with the relevant Fleet

Reviewer: ```suggestion
optional `--application` switch to the command with the relevant fleet
```

name:

```shell
$ cd ~/balenaProduction
$ balena device move 94095f8 --application altFleet
94095f8 was moved to altFleet
```

A call of `balena device` specifying the UUID of the moved device will now

> show it is owned by the specified Fleet:

Reviewer: ```suggestion
show it is owned by the specified fleet:
```

The Supervisor on the device will remove any previously running services, as
well as their images and any volumes associated with them and download the

> images associated with the new Fleet before starting them.

Reviewer: ```suggestion
images associated with the new fleet before starting them.
```

First, provision a device using a development image. You can do this using
the balenaCloud dashboard and downloading a 'Development' edition image from

> the `cliFleet` Fleet page. Provision your balenaFin with this image

Reviewer: ```suggestion
the `cliFleet` fleet summary page. Provision your balenaFin with this image
```

$ cd ~/balenaProduction
$ balena devices

> ID      UUID    DEVICE NAME      DEVICE TYPE     APPLICATION NAME STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL

Reviewer: ```suggestion
ID      UUID    DEVICE NAME      DEVICE TYPE     FLEET NAME STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL
```

STATUS:             idle
IS ONLINE:          true
IP ADDRESS:         192.168.1.171

> APPLICATION NAME:   cliFleet

Reviewer: ```suggestion
FLEET NAME:   cliFleet
```

We can now join a different balena environment by using balena CLI to login
to it. As we previously did this for our staging environment, we can simply
use the data and tokens we saved for this by changing directories and using

> the other environment information. We'll now create a new Fleet on the

Reviewer: ```suggestion
the other environment information. We'll now create a new fleet on the
```

staging environment to move the device to:

```shell
$ cd ~/balenaStaging

> $ balena app create stagingCliFleet --type fincm3
> Application created: stagingCliFleet (fincm3, id 97654)

Reviewer: ```suggestion
$ balena fleet create stagingCliFleet --type fincm3
Fleet created: stagingCliFleet (fincm3, id 97654)
```

$ cd ~/balenaStaging
$ balena devices

> ID     UUID    DEVICE NAME      DEVICE TYPE  APPLICATION NAME  STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL

Reviewer: ```suggestion
ID     UUID    DEVICE NAME      DEVICE TYPE  FLEET NAME  STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL
```

876542 3456789 purple-snowflake fincm3       stagingCliFleet     Idle   true      9.15.7             balenaOS 2.38.0+rev1 https://dashboard.balena-staging.com/devices/3456789345678934567893456789/summary
```

As can be seen, it's been given a new name and UUID.

> If we hadn't specified the Fleet to join, we would have seen an
> interactive list of all the Fleets on the staging environment:

Reviewer: ```suggestion
If we hadn't specified the fleet to join, we would have seen an
interactive list of all the fleets on the staging environment:
```

```shell
$ cd ~/balenaStaging
$ balena join 192.168.1.171
? Select application
❯ heds/artik530
  heds/orangepi
  heds/stagingclifleet
  heds/switchapp
  heds/vpn-changed
```

> Note that the Fleets listed are not all of the `fincm3` device type, but

Reviewer: ```suggestion
Note that the fleets listed are not all of the `fincm3` device type, but
```

share a common architecture, so this could be `aarch64`, `armv7hf` or `armv6`

> for the Fin. You should be very careful when selecting a Fleet of a

Reviewer: ```suggestion
for the Fin. You should be very careful when selecting a fleet of a
```

A downloaded balenaOS image via balena CLI is unconfigured, so to allow a device
to use it as a provisioning image we need to specify, at a minimum, which

> Fleet the device should be associated with.

Reviewer: ```suggestion
fleet the device should be associated with.
```

There are a few ways to achieve this. The simplest is to configure it

> interactively, passing either a Fleet name or device UUID so that

Reviewer: ```suggestion
interactively, passing either a fleet name or device UUID so that
```

> the relevant Fleet can be determined:

Reviewer: ```suggestion
the relevant fleet can be determined:
```

configuration into a bare OS image.

This allows multiple configs to be generated, for example for different

> Fleets, and then using copies of a bare OS image of a particular version

Reviewer: ```suggestion
fleets, and then using copies of a bare OS image of a particular version
```

to prepare images for each of those apps.

You can generate independent configuration files using the
`balena config generate` command. A mandatory OS version must be passed to this

> command, with either an Fleet name or device UUID. It will interactively

Reviewer: ```suggestion
command, with either an fleet name or device UUID. It will interactively
```

> we'll get some debug logging. So let's use `balena env` to do this:

Reviewer: ```suggestion
we'll get some debug logging (If you don't see this message ensure you are pushing the code from this masterclass, and not `balena-cli-masterclass`!). So let's use `balena env` to do this:
```

Reviewer: Adding this change from https://github.com/balena-io/balena-cli-advanced-masterclass/pull/30/files

```shell
$ balena env add LOG_DEBUG true --application cliFleet
```

> Now go to the dashboard for the `cliFleet` Fleet, and select 'Environment

Reviewer: ```suggestion
Now go to the dashboard for the `cliFleet` fleet, and select '
```

```shell
$ balena env rm 654321 --yes

> $ balena envs --application cliFleet

Reviewer: ```suggestion
$ balena envs --fleet cliFleet
```

We're back to no debug logs again!

The balena CLI also allows you to rename environment variables, as well as set

> and remove them for specific devices instead of an entire Fleet.

Reviewer: ```suggestion
and remove them for specific devices instead of an entire fleet.
```

shipping it's only half the story.

Preregistering a device allows the creation of individually registered devices

> to a Fleet before those devices are ever physically powered on or

Reviewer: ```suggestion
to a fleet before those devices are ever physically powered on or
```

`/Work/images` to Docker Desktop's File Sharing resources).

**Important Note:** Currently, Docker for Windows and Docker for Mac only

> 	ship with the 'overlay2' storage driver. This means that any Fleet

Reviewer: ```suggestion
	ship with the 'overlay2' storage driver. This means that any fleet
```

	image that does not use 'overlay2' as the storage driver (including those
	for the balena Fin), can not be preloaded under these host platforms.

> We'll take the current `cliFleet` Fleet and preload it into the same

Reviewer: ```suggestion
We'll take the current `cliFleet` fleet and preload it into the same
```

`balena-fin-image.img` image we earlier configured and provisioned our device
with. The `balena preload` command has a large number of switch options for
catering with different situations, including the ability to use a particular
release. However, for this exercise, we'll simply use the latest version of
the code we previously pushed to `cliFleet`. In a terminal, execute the following:

```shell

> $ balena preload balena-fin-image.img --app cliFleet --commit latest

Reviewer: ```suggestion
$ balena preload balena-fin-image.img --fleet cliFleet --commit latest
```

Building Docker preloader image. [========================] 100%
/ Creating preloader container
\ Starting preloader container

> - Fetching application cliFleet

Reviewer: ```suggestion
- Fetching fleet cliFleet
```

image to the balenaFin.

We'll verify that we're registering a device by first listing the devices
currently associated with the `cliFleet` Fleet:

```shell

> $ balena devices --application cliFleet

Reviewer: ```suggestion
$ balena devices --fleet cliFleet
```

now:

```shell
$ balena device register cliFleet --uuid 6053dab8dc4721ed288c8dfc79e52967
BalenaRequestError: Request error: "uuid" must be unique.
```

> Now we'll look at the device list for the Fleet again:

Reviewer: ```suggestion
Now we'll look at the device list for the fleet again:
```

```shell

> $ balena devices --application cliFleet

Reviewer: ```suggestion
$ balena devices --fleet cliFleet
```

> ID      UUID    DEVICE NAME      DEVICE TYPE     APPLICATION NAME STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL

Reviewer: ```suggestion
ID      UUID    DEVICE NAME      DEVICE TYPE     FLEET NAME STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL
```

exercise:

```shell

> $ balena config generate --app cliFleet --version 2.38.0+rev1 --device 6053dab8dc4721ed288c8dfc79e52967 --network ethernet --appUpdatePollInterval 10 --output config.json

Reviewer: ```suggestion
$ balena config generate --fleet cliFleet --version 2.38.0+rev1 --device 6053dab8dc4721ed288c8dfc79e52967 --network ethernet --appUpdatePollInterval 10 --output config.json
```

preregistered device come online:

```shell

> $ balena devices --application cliFleet

Reviewer: ```suggestion
$ balena devices --fleet cliFleet
```

> ID      UUID    DEVICE NAME      DEVICE TYPE     APPLICATION NAME STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL

Reviewer: ```suggestion
ID      UUID    DEVICE NAME      DEVICE TYPE     FLEET NAME STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL
```

1696632 f1dd777 empty-sun        fincm3          cliTest          Idle   false     10.2.2             balenaOS 2.41.0+rev4 https://dashboard.balena-cloud.com/devices/12345678901234567890123456789012/summary
1699866 93b40bc late-sunset      fincm3          cliFleet           Idle   true      10.2.2             balenaOS 2.41.0+rev4 https://dashboard.balena-cloud.com/devices/93b40bc440cfccc6d45fa8db4a777a06/summary
```

> As before, the application will also start immediately as it was preloaded.

Reviewer: ```suggestion
As before, the release(s) will also start immediately as it was preloaded.
```

verify this with:

```shell

> $ balena devices --application cliFleet

Reviewer: ```suggestion
$ balena devices --fleet cliFleet
```

> ID      UUID    DEVICE NAME      DEVICE TYPE  APPLICATION NAME STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL

Reviewer: ```suggestion
ID      UUID    DEVICE NAME      DEVICE TYPE  FLEET NAME STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL
```

Finally, let's run `balena devices` again to see the new version of the device:

```shell

> $ balena devices --application cliFleet

Reviewer: ```suggestion
$ balena devices --fleet cliFleet
```

> ID      UUID    DEVICE NAME      DEVICE TYPE  APPLICATION NAME STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL

Reviewer: ```suggestion
ID      UUID    DEVICE NAME      DEVICE TYPE  FLEET NAME STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL
```

enough with balena CLI to:

* Switch between balena environments, as well as move devices between them

>     as well as different Fleets

Reviewer: ```suggestion
    as well as different fleets
```

* Download, configure and provision device images using balena CLI

> * Modify environment and configuration variables, both Fleet and device

Reviewer: ```suggestion
* Modify environment and configuration variables, both fleet and device
```
